simplify profile text output by displaying only top-level ops statistics

YazhiGao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42262 simplify profile text output by displaying only top-level ops statistics**

Differential Revision: [D22830328](https://our.internmc.facebook.com/intern/diff/D22830328)

Commit:
simplify profile text output by displaying only top-level ops statistics

[ghstack-poisoned]

Commit:
Update on "simplify profile text output by displaying only top-level ops statistics"




[ghstack-poisoned]

Commit:
Update on "simplify profile text output by displaying only top-level ops statistics"


Differential Revision: [D22904383](https://our.internmc.facebook.com/intern/diff/D22904383)

[ghstack-poisoned]

YazhiGao commented:
This is a problem revisit of user trying to see only top-level ops time attributions in terms of all the ops(some vaguely named to be differentiated) nested down to cpp/cuda ops.
quoting some of the internal user feedback here in some discussion posts:
```
Folks, I appreciate the insights. However, here is my feedback as a user of the profiler.
...As a user, I most often do not care what implementation I have called underneath. I only care to see the top operator perf. Caffe2 provides detailed and summarized output as part of its profiler and perhaps we should follow such a strategy as well. For instance look at the 6 outputs for the embeddingbag operator, it's impossible to read (and this is a single operator not the entire model!):
1. EmbeddingBagBackward 1.55% 419.895us 17.68% 4.779ms 398.248us NaN 0.000us 0.000us 12
2. _embedding_bag_backward 1.70% 458.349us 16.13% 4.359ms 363.257us NaN 0.000us 0.000us 12
add_ 11.52% 3.115ms 11.52% 3.115ms 59.905us NaN 0.000us 0.000us 52
...
4. embedding_bag 0.40% 107.873us 4.80% 1.298ms 108.151us NaN 0.000us 0.000us 12
...
6. _embedding_bag 4.28% 1.157ms 4.28% 1.157ms 96.448us NaN 0.000us 0.000us
```

```
==================================================================================================================================================================================
TEST
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
Name                           Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls  Input Shapes                                   
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
aten::add_                     3.61%            462.489us        3.61%            462.489us        462.489us        1                [[3, 20], [3, 20], []]                         
aten::slice                    1.95%            249.571us        1.95%            250.018us        250.018us        1                [[3, 80], [], [], [], []]                      
aten::lstm                     1.89%            242.534us        22.41%           2.872ms          2.872ms          1                [[5, 3, 10], [], [], [], [], [], [], [], []]   
aten::lstm                     1.68%            215.852us        18.18%           2.330ms          2.330ms          1                [[5, 3, 10], [], [], [], [], [], [], [], []]   
aten::lstm                     1.68%            215.767us        18.49%           2.370ms          2.370ms          1                [[5, 3, 10], [], [], [], [], [], [], [], []]   
aten::lstm                     1.60%            205.014us        20.15%           2.582ms          2.582ms          1                [[5, 3, 10], [], [], [], [], [], [], [], []]   
aten::lstm                     1.55%            198.213us        18.53%           2.375ms          2.375ms          1                [[5, 3, 10], [], [], [], [], [], [], [], []]   
aten::addmm                    0.95%            122.359us        1.01%            129.857us        129.857us        1                [[80], [3, 20], [20, 80], [], []]              
aten::stack                    0.29%            36.745us         0.63%            80.179us         80.179us         1                [[], []]                                       
aten::add_                     0.28%            35.694us         0.28%            35.694us         35.694us         1                [[3, 20], [3, 20], []]                         
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
Self CPU time total: 12.817ms

-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
Name                           Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls  Input Shapes                                   
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
aten::mul                      11.45%           1.467ms          12.88%           1.651ms          11.006us         150              [[3, 20], [3, 20]]                             
aten::lstm                     8.41%            1.077ms          97.76%           12.529ms         2.506ms          5                [[5, 3, 10], [], [], [], [], [], [], [], []]   
aten::addmm                    7.65%            979.982us        11.38%           1.459ms          29.182us         50               [[80], [3, 20], [20, 80], [], []]              
aten::sigmoid_                 6.78%            869.295us        9.74%            1.249ms          8.327us          150              [[3, 20]]                                      
aten::add_                     5.82%            745.801us        5.82%            745.801us        14.916us         50               [[3, 20], [3, 20], []]                         
aten::slice                    5.58%            715.532us        6.61%            847.445us        4.237us          200              [[3, 80], [], [], [], []]                      
aten::unsafe_split             4.24%            544.015us        13.25%           1.698ms          33.957us         50               [[3, 80], [], []]                              
aten::tanh                     3.11%            398.881us        6.05%            775.024us        15.500us         50               [[3, 20]]                                      
aten::empty                    3.04%            389.055us        3.04%            389.055us        1.319us          295              [[], [], [], [], [], []]                       
aten::sigmoid                  2.96%            379.686us        2.96%            379.686us        2.531us          150              [[3, 20], [3, 20]]                             
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
Self CPU time total: 12.817ms

==================================================================================================================================================================================
TEST
==================================================================================================================================================================================
This report only display top-level ops statistics
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
Name                           Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls  Input Shapes                                   
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
aten::lstm                     1.89%            242.534us        22.41%           2.872ms          2.872ms          1                [[5, 3, 10], [], [], [], [], [], [], [], []]   
aten::lstm                     1.68%            215.852us        18.18%           2.330ms          2.330ms          1                [[5, 3, 10], [], [], [], [], [], [], [], []]   
aten::lstm                     1.68%            215.767us        18.49%           2.370ms          2.370ms          1                [[5, 3, 10], [], [], [], [], [], [], [], []]   
aten::lstm                     1.60%            205.014us        20.15%           2.582ms          2.582ms          1                [[5, 3, 10], [], [], [], [], [], [], [], []]   
aten::lstm                     1.55%            198.213us        18.53%           2.375ms          2.375ms          1                [[5, 3, 10], [], [], [], [], [], [], [], []]   
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
Self CPU time total: 12.817ms

==================================================================================================================================================================================
This report only display top-level ops statistics
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
Name                           Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls  Input Shapes                                   
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
aten::lstm                     8.41%            1.077ms          97.76%           12.529ms         2.506ms          5                [[5, 3, 10], [], [], [], [], [], [], [], []]   
-----------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------------------------------------  
Self CPU time total: 12.817ms

Total time based on python measurements:  13.206ms
CPU time measurement python side overhead: 3.03%
```

Commit:
Update on "simplify profile text output by displaying only top-level ops statistics"


Differential Revision: [D22904383](https://our.internmc.facebook.com/intern/diff/D22904383)

[ghstack-poisoned]

Commit:
Update on "simplify profile text output by displaying only top-level ops statistics"


Differential Revision: [D22904383](https://our.internmc.facebook.com/intern/diff/D22904383)

[ghstack-poisoned]

YazhiGao commented:
ping @ilia-cher for review

ilia-cher commented on torch/autograd/profiler.py:
```diff
@@ -114,6 +120,12 @@ def table(self, sort_by=None, row_limit=100, header=None):
+            top_level_events_only(bool, optional): Boolean flag to determine the
+                selection of events to display. If true, the profiler will only
+                display events at top level like top-level invocation of python
+                `lstm`, python `add` or other functions, nested events like cpu/cuda
+                ops events like `EmbeddingBagBackward`, `_embedding_bag_backward` are
```
I would remove explicit examples here, and also it's not clear why e.g. EmbeddingBagBackward should not be printed

ilia-cher commented on torch/autograd/profiler.py:
```diff
@@ -231,6 +244,19 @@ def total_average(self):
         total_stat.key = 'Total'
         return total_stat
 
+    def without_nested_events(self):
```
I don't see how this API is used, since this is a public API let's keep it to a minimum and add things only when necessary

Commit:
Update on "simplify profile text output by displaying only top-level ops statistics"


Differential Revision: [D22830328](https://our.internmc.facebook.com/intern/diff/D22830328)

[ghstack-poisoned]

ilia-cher reviewed (approved):
LG!
